xcm/xcm-emulator/README.md

```diff
@@ -0,0 +1,110 @@
+# xcm-emulator
+
+XCM-Emulator is a tool to emulate XCM program execution using
+pre-configured runtimes, including those used to run on live
+networks, such as Kusama, Polkadot, Statemine et cetera.
+This allows for testing cross-chain message passing, verifying
+outcomes, weights and side-effects. It is faster than spinning up
+a zombienet and as all the chains are in one process debugging using Clion is easy.
+
+## Limitations
+
+As the channels are mocked, using xcm-emulator tests to test
+channel setup would not be appropriate.
```

Comment:
It’s using `*Sink` to facilitate message passing as opposed to `xcm-simulator` that is using mocks. Opening of channels is not required though.

Reply:
> Opening of channels is not required though.

what channels? e.g. hrmp channel needs to be opened, else parachain's XcmRouter->XcmpQueue sending messages wont work.

If "As the channels are mocked" mean ump/dmp/hrmp, I wouldnt say they are mocked. Just to read/write mesasges from these queues is orchestrated directly here with some thread_local queues without `relay-chain-interface / relay-chain-inprocess-interface` code, so lets say it uses testing/simplified transport layer between relaychain and parachain.

Reply:
> Could rewrite as: "As the messages do not physically go through the same messaging infrastructure there is some code that is not being tested compared to using slower E2E tests. In future it may be possible to run these XCM emulated tests as E2E tests (without changes)."

Sounds perfect. It's indeed not 100% the same, but not fully mocked either.

Reply:
Fixed [here](https://github.com/paritytech/cumulus/pull/2447)